Clarify failure message for quaffing !invis while haloed (Yermak, #11104)

jmbjr opened the pull request:
See: https://crawl.develz.org/mantis/view.php?id=11104

Currently, self-haloing prevents !invis, even when umbraed.

The issue is, the msg to the player is vague and doesn't clarify that the reason the player can't go invisible at all is due to their weapon (EOS) or their Divine Halo (TSO or Wu Jian). Additionally, since the player is actually able to quaff !invis when non-self-haloed haloed (angels, sunshine, etc) + umbraed, it could be confusing and feel inconsistent. 

With this fix, the actual reason for preventing quaffing !invis is provided to the player.

I don't know if I'm using ASSERT appropriately, but my concern was a future halo source that wouldn't get properly handled by this message.  I'm assuming there may be a better way to do what I'm trying to do.

I also briefly considered just allowing quaffing !invis always if you were haloed (any source) + umbraed, but the code seems to intentionally make the distinction between self-halos and simply being backlit by an external halo source, so I didn't change the behavior.

jmbjr commented:
note 1:  you could argue that the existing message, "Your halo.." already indicates you are self-haloed, but adding the actual reason, the weapon or Divine Halo, seems much more clear.

note 2: I didn't know if it was OK to capitalise Divine Halo as a way to reinforce it's due to a god effect.

amalloy commented on crawl-ref/source/player.cc:
```diff
@@ -4745,11 +4745,21 @@ void dec_channel_player(int delay)
+            reason = (reason == "" ? "Divine Halo" : reason + " and Divine Halo");
+        }
+
+        ASSERT(reason != "");
```
`!reason.empty()`

jmbjr replied:
I take it this means this is a proper use of ASSERT?  (and doh on the .empty(), ty)

amalloy commented on crawl-ref/source/player.cc:
```diff
@@ -4745,11 +4745,21 @@ void dec_channel_player(int delay)
+        if (you.attribute[ATTR_HEAVENLY_STORM] > 0 ||
+            you.religion == GOD_SHINING_ONE)
+        {
+            reason = (reason == "" ? "Divine Halo" : reason + " and Divine Halo");
```
`.empty()`.

amalloy replied:
Also I think capitalising it is wrong. It's not an ability named "Divine Halo" (right?), it's a halo which is divine.

amalloy commented on crawl-ref/source/player.cc:
```diff
@@ -4745,11 +4745,21 @@ void dec_channel_player(int delay)
+        }
+
+        ASSERT(reason != "");
+        msg = "Your " + reason + " prevents invisibility.";
```
Should use "prevent" if there is more than one reason.

jmbjr replied:
this actually is irrelevant now after the changes to msg upstream.

amalloy commented on crawl-ref/source/player.cc:
```diff
@@ -4745,11 +4745,21 @@ void dec_channel_player(int delay)
     {
-        msg = "Your halo prevents invisibility.";
+        if (player_equip_unrand(UNRAND_EOS))
+            reason = "weapon";
```
I would say "weapon's halo" or something like that, to explain why the weapon makes invisibility impossible. Maybe even change the "your ____ prevents invisibility" style to "Your weapon shines too brightly to become invisible".

jmbjr replied:
Great idea!

jmbjr commented:
I updated the messages. I really liked the suggestion to add a bit more flavor to it and vastly prefer the new messages now.

all other suggestions also addressed/fixed.

amalloy commented on crawl-ref/source/player.cc:
```diff
@@ -4745,11 +4745,24 @@ void dec_channel_player(int delay)
     {
-        msg = "Your halo prevents invisibility.";
+        if (player_equip_unrand(UNRAND_EOS))
+            reason = "Your weapon shines too brightly";
```
It feels unsatisfying to me to set this reason and then perhaps overwrite it later. In particular, this approach only works because you currently have exactly two halo sources. If there were more, you wouldn't be able to know which halo source had previously been set.

Better, I think, is to have non-overlapping `if` clauses, like this:

```
bool divine = (...);
bool weapon = (...);

if (divine && weapon)
  ...
else if (divine)
  ...
else if (weapon)
  ...
else
  die("haloed by an unknown source");
```

jmbjr commented:
Updated based on latest input.  I agree that it seems to flow better now.

I didn't know about die().  Is that a better way to handle the situation that I was concerned with on if a new halo source was added and not handled here? I removed the ASSERT(!message.empty()), but I wasn't sure if that was OK or if I should have both the die() and ASSERT.

amalloy commented:
IMO check with @neilmoore about whether die is good here. I think it provides the same information as assert but without the conditional, in which case it's perfect here.

amalloy commented:
I just looked at the source and indeed `die` is fine. It's the right thing to use when you want to crash because you've wandered into an impossible codepath as opposed to because a variable has taken on an impossible value. I approve these changes and will merge them in if you're done.

jmbjr commented:
I am done. Thanks!
